src/NHibernate/Dialect/Dialect.cs

```diff
@@ -2397,9 +2399,9 @@ public virtual ISQLExceptionConverter BuildSQLExceptionConverter()
 
 		internal static string GetDriverName<TFallback>(string preferredName) where TFallback : IDriver
 		{
-			return ReflectHelper.ClassForFullNameOrNull(preferredName) == null
-				? typeof(TFallback).FullName
-				: preferredName;
+			return GetDriverNameMemoize.GetOrAdd(
+				preferredName,
+				s => ReflectHelper.ClassForFullNameOrNull(s) == null ? typeof(TFallback).FullName : s);
```

Comment:
Strictly speaking, incorrect. Currently, `TFallback` is always the same for a given `preferredName`, but nothing guarantees this. This is why my proposed code was not caching the function result but only the preferred type resolution.